keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+
+The `kubeadm join --master` workflow will be implemented as an extension of the
+existing `kubeadm join` flow.
+
+`kubeadm join --master` will accept an additional parameter, that is the apiserver advertise
+address of the joining node; as details in following paragraphs, the value assigned to
+this parameter depends on the user choice between a dynamic bootstrap workflow or a static
+bootstrap workflow.
+
+The updated join workflow will be the following:
+
+1. Discovery cluster info [No changes to this step]
+
+   > NB This step waits for a first instance of the kube-apiserver to become ready
+   > (the bootstrap master); And thus it acts as embedded mechanism for handling the sequence
+   > `kubeadm init` and `kubeadm join` actions in case of parallel node creation.
+
+2. In case of `join --master` [New step]
+
+   1. Using the bootstrap token as identity, read the `kubeadm-config` configMap
+      in `kube-system` namespace.
+
+      > This requires to grant access to the above configMap for
+      > `system:bootstrappers` group.
+
+   2. Check if the cluster is ready for joining a new master node:
+
+      a. Check if the cluster has a stable `controlplaneAddress`
+      a. Check if the cluster uses an external etcd
+      a. Checks if the mandatory certificates exists on the file system
+
+   3. Prepare the node for joining as a master node:
+
+      a. Create missing certificates (in any).
+         > please note that by creating missing certificates kubeadm can adapt seamlessly
+         > to a dynamic workflow or to a static workflow (and to apiserver advertise address
+         > of the joining node). see following paragraphs for more details for additional info.
+
+      a. In case of control plane deployed as static pods, create related kubeconfig files
+      and static pod manifests.
+
+      > see "Strategies for deploying control plane components" paragraph
+      > for additional info about this step.
+
+   4. Create the admin.conf kubeconfig file
+
+3. Executes the kubelet TLS bootstrap process [No changes to this step]:
+
+4. In case of `join --master` [New step]
+
+   1. Apply master taint and label to the node.
+
+      > This action is executed using the admin.conf identity created above;
+
+   2 Update the `kubeadm-config` configMap with the information about the new master node
+
+#### dynamic workflow (advertise-address == `controlplaneAddress`)
+
+There are many ways to configure an highly available cluster.
+
+Among them, the approach best suited for a dynamic bootstrap workflow requires the
+user to set the `--apiserver-advertise-address` of each master, including the bootstrap master
+itself, equal to the `controlplaneAddress` endpoint provided during kubeadm init
+(the IP/DNS of the external load balancer).
+
+By using the same advertise address for all the IP masters, `kubeadm init` can create
+a unique API server serving certificate that could be shared across many masters nodes;
```

Comment:
??
I think the certificate should be **always** be unique per-master but always have `controlPlaneAddress` in the SANs if specified.

Reply:
This is not necessary in case of advertise-address == `controlplaneAddress`, but nothing prevents you to do so.

As stated in the note, kubeadm will act as you suggest, creating unique certificates per-master unless the user intentionally decide to distribute the same certificate everywhere (that is something we will not recommend in the user guide)